### ISSUE TITLE
const in a use statement gets wrongly parsed as T_STRING instead of T_CONST

### ISSUE THREAD

**Issue by ExplodingCabbage** [squizlabs/PHP_CodeSniffer#1385]
PHPCS wrongly parses files like this:

```php
<?php
use const X\Y;
```

The `const` should be a `T_CONST` but is treated as a `T_STRING` instead.

Example when running `PHP_CodeSniffer/scripts/phpcs -vv test.php` against the script above using the current HEAD of master (1ad86bf7e5486def140344fd501b9e07b0c96a72), with the problematic output bolded:

<pre><code>Processing ruleset /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/ruleset.xml
	Adding sniff files from &quot;/.../PEAR/Sniffs/&quot; directory
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Formatting/MultiLineAssignmentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Functions/FunctionCallSignatureSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Functions/ValidDefaultValueSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Functions/FunctionDeclarationSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/WhiteSpace/ScopeClosingBraceSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/WhiteSpace/ScopeIndentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/WhiteSpace/ObjectOperatorIndentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Files/IncludingFileSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Classes/ClassDeclarationSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Commenting/FileCommentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Commenting/FunctionCommentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Commenting/InlineCommentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/Commenting/ClassCommentSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/ControlStructures/ControlSignatureSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/ControlStructures/MultiLineConditionSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/NamingConventions/ValidClassNameSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/NamingConventions/ValidFunctionNameSniff.php
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/PEAR/Sniffs/NamingConventions/ValidVariableNameSniff.php
	Processing rule &quot;Generic.Functions.FunctionCallArgumentSpacing&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/Functions/FunctionCallArgumentSpacingSniff.php
	Processing rule &quot;Generic.NamingConventions.UpperCaseConstantName&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/NamingConventions/UpperCaseConstantNameSniff.php
	Processing rule &quot;Generic.PHP.LowerCaseConstant&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/PHP/LowerCaseConstantSniff.php
	Processing rule &quot;Generic.PHP.DisallowShortOpenTag&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/PHP/DisallowShortOpenTagSniff.php
	Processing rule &quot;Generic.WhiteSpace.DisallowTabIndent&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/WhiteSpace/DisallowTabIndentSniff.php
	Processing rule &quot;Generic.Commenting.DocComment&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/Commenting/DocCommentSniff.php
	Processing rule &quot;Generic.Files.LineLength&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/Files/LineLengthSniff.php
		=&gt; property &quot;lineLimit&quot; set to &quot;85&quot;
		=&gt; property &quot;absoluteLineLimit&quot; set to &quot;0&quot;
	Processing rule &quot;Generic.Files.LineEndings&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/Files/LineEndingsSniff.php
		=&gt; property &quot;eolChar&quot; set to &quot;\n&quot;
	Processing rule &quot;Generic.Functions.FunctionCallArgumentSpacing.TooMuchSpaceAfterComma&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/Functions/FunctionCallArgumentSpacingSniff.php
		=&gt; severity set to 0
	Processing rule &quot;Generic.ControlStructures.InlineControlStructure&quot;
		=&gt; /home/mark/PHP_CodeSniffer/CodeSniffer/Standards/Generic/Sniffs/ControlStructures/InlineControlStructureSniff.php
		=&gt; property &quot;error&quot; set to &quot;false&quot;
=&gt; Ruleset processing complete; included 27 sniffs and excluded 0
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_DO =&gt; do
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [2]: T_WHILE =&gt; while
	Process token [3]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  4 : T_OPEN_PARENTHESIS =&gt; (
	Process token [5]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token  1 : T_SEMICOLON =&gt; ;
	Process token [2]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHILE =&gt; while
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_PARENTHESIS =&gt; (
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_FOR =&gt; for
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_PARENTHESIS =&gt; (
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_IF =&gt; if
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_PARENTHESIS =&gt; (
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_FOREACH =&gt; foreach
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_PARENTHESIS =&gt; (
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token  1 : T_CLOSE_CURLY_BRACKET =&gt; }
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [3]: T_ELSE =&gt; else
	Process token [4]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [5]: T_IF =&gt; if
	Process token [6]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  7 : T_OPEN_PARENTHESIS =&gt; (
	Process token [8]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token  1 : T_CLOSE_CURLY_BRACKET =&gt; }
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [3]: T_ELSEIF =&gt; elseif
	Process token [4]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  5 : T_OPEN_PARENTHESIS =&gt; (
	Process token [6]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  2 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [3]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token  1 : T_CLOSE_CURLY_BRACKET =&gt; }
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [3]: T_ELSE =&gt; else
	Process token [4]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  5 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [6]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m·[0m
	Process token [1]: T_DO =&gt; do
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token  3 : T_OPEN_CURLY_BRACKET =&gt; {
	Process token [4]: T_CLOSE_TAG =&gt; ?&gt;
	*** END PHP TOKENIZING ***
Creating file list... DONE (1 files in queue)
Changing into directory /home/mark
Processing test.php 
	*** START PHP TOKENIZING ***
	Process token [0]: T_OPEN_TAG =&gt; &lt;?php[30;1m\n[0m
	Process token [1]: T_USE =&gt; use
	Process token [2]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [3]: T_CONST =&gt; const
	Process token [4]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token [5]: T_STRING =&gt; X
	Process token [6]: T_NS_SEPARATOR =&gt; \
	Process token [7]: T_STRING =&gt; Y
	Process token  8 : T_SEMICOLON =&gt; ;
	Process token [9]: T_WHITESPACE =&gt; [30;1m\n[0m
	*** END PHP TOKENIZING ***
	*** START TOKEN MAP ***
	*** END TOKEN MAP ***
	*** START SCOPE MAP ***
	Start scope map at 1:T_USE =&gt; use
	=&gt; Begin scope map recursion at token 1 with depth 1
	Process token 2 on line 2 []: T_WHITESPACE =&gt; [30;1m·[0m
	Process token 3 on line 2 []: <b>T_STRING =&gt; const</b>
	Process token 4 on line 2 []: T_WHITESPACE =&gt; [30;1m·[0m
	Process token 5 on line 2 []: T_STRING =&gt; X
	Process token 6 on line 2 []: T_NS_SEPARATOR =&gt; \
	Process token 7 on line 2 []: T_STRING =&gt; Y
	Process token 8 on line 2 []: T_SEMICOLON =&gt; ;
	=&gt; Found semicolon before scope opener for 1:T_USE, bailing
	*** END SCOPE MAP ***
	*** START LEVEL MAP ***
	Process token 0 on line 1 [col:1;len:5;lvl:0;]: T_OPEN_TAG =&gt; &lt;?php[30;1m\n[0m
	Process token 1 on line 2 [col:1;len:3;lvl:0;]: T_USE =&gt; use
	Process token 2 on line 2 [col:4;len:1;lvl:0;]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token 3 on line 2 [col:5;len:5;lvl:0;]: <b>T_STRING =&gt; const</b>
	Process token 4 on line 2 [col:10;len:1;lvl:0;]: T_WHITESPACE =&gt; [30;1m·[0m
	Process token 5 on line 2 [col:11;len:1;lvl:0;]: T_STRING =&gt; X
	Process token 6 on line 2 [col:12;len:1;lvl:0;]: T_NS_SEPARATOR =&gt; \
	Process token 7 on line 2 [col:13;len:1;lvl:0;]: T_STRING =&gt; Y
	Process token 8 on line 2 [col:14;len:1;lvl:0;]: T_SEMICOLON =&gt; ;
	Process token 9 on line 2 [col:15;len:0;lvl:0;]: T_WHITESPACE =&gt; [30;1m\n[0m
	*** END LEVEL MAP ***
	*** START ADDITIONAL PHP PROCESSING ***
	*** END ADDITIONAL PHP PROCESSING ***
[PHP =&gt; 10 tokens in 2 lines]... 
DONE in 1ms (1 errors, 0 warnings)

FILE: /home/mark/test.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 2 | ERROR | Missing file doc comment
----------------------------------------------------------------------

Time: 20ms; Memory: 4Mb
</code></pre>

I admit that I'm very confused about how this doesn't break any tests, since the `UseDeclarationSniff` checks for `T_CONST` in this context (and has a test for it). This *is* a bug, right?

This is blocking https://github.com/squizlabs/PHP_CodeSniffer/pull/1106#issuecomment-285900576 since I don't want to write code that relies upon incorrect tokenizer output.

**Comment by ExplodingCabbage** [comment]
Oh, this was on PHP 7.0.15 on Ubuntu, just in case it matters.

**Comment by gsherwood** [comment]
This is intentional and was introduced as a fix for bug #364

The intention here is that these are not function or const definitions, and so should be ignored by sniffs that are checking for them. This would include things like naming conventions, which should be checked where the const was originally defined.

If you are looking for used functions and constants, you'll need to look for T_USE and then see what type of construct is being used by looking for the next string.
